src/olympia/activity/models.py

```diff
@@ -49,3 +70,412 @@ class ActivityLogEmails(ModelBase):
+                review = self.f(u'<a href="{0}">{1}</a>',
+                                arg.get_url_path(), _('Review'))
+                arguments.remove(arg)
+            if isinstance(arg, Version) and not version:
+                text = _('Version {0}')
+                if arg.channel == amo.RELEASE_CHANNEL_LISTED:
+                    version = self.f(u'<a href="{1}">%s</a>' % text,
+                                     arg.version, arg.get_url_path())
+                else:
+                    version = self.f(text, arg.version)
+                arguments.remove(arg)
+            if isinstance(arg, Collection) and not collection:
+                collection = self.f(u'<a href="{0}">{1}</a>',
+                                    arg.get_url_path(), arg.name)
+                arguments.remove(arg)
+            if isinstance(arg, Tag) and not tag:
+                if arg.can_reverse():
+                    tag = self.f(u'<a href="{0}">{1}</a>',
+                                 arg.get_url_path(), arg.tag_text)
+                else:
+                    tag = self.f('{0}', arg.tag_text)
+            if isinstance(arg, Group) and not group:
+                group = arg.name
+                arguments.remove(arg)
+            if isinstance(arg, File) and not file_:
+                validation = 'passed'
+                if self.action in (
+                        amo.LOG.BETA_SIGNED.id,
+                        amo.LOG.BETA_SIGNED_VALIDATION_FAILED.id,
+                        amo.LOG.UNLISTED_SIGNED.id,
+                        amo.LOG.UNLISTED_SIGNED_VALIDATION_FAILED.id):
+                    validation = 'ignored'
+
+                file_ = self.f(u'<a href="{0}">{1}</a> (validation {2})',
+                               reverse('files.list', args=[arg.pk]),
+                               arg.filename,
+                               validation)
+                arguments.remove(arg)
+
+        user = user_link(self.user)
+
+        try:
+            kw = dict(addon=addon, review=review, version=version,
+                      collection=collection, tag=tag, user=user, group=group,
+                      file=file_)
+            return self.f(format, *arguments, **kw)
+        except (AttributeError, KeyError, IndexError):
+            log.warning('%d contains garbage data' % (self.id or 0))
+            return 'Something magical happened.'
+
+    def __unicode__(self):
+        return self.to_string()
+
+    def __html__(self):
+        return self
+
+    @classmethod
+    def create(cls, action, *args, **kw):
+        """
+        e.g. ActivityLog.create(amo.LOG.CREATE_ADDON, []),
+             ActivityLog.create(amo.LOG.ADD_FILE_TO_VERSION, file, version)
+        In case of circular import you can use `olympia.activity.log_create()`
+        """
+        from olympia import core
+
+        user = kw.get('user', core.get_user())
+
+        if not user:
+            olympia.core.logger.getLogger('z.amo').warning(
```

Comment:
Another follow-up task, worth moving that `getLogger` call above and use `activity` (or `z.activity` ?) instead of `z.amo`, maybe ? Need to double-check logging config